ecosystem/indexer-grpc/indexer-grpc-table-info/src/internal_indexer_db_service.rs

```diff
@@ -166,31 +166,32 @@ impl InternalIndexerDBService {
 
     pub async fn run(&mut self, node_config: &NodeConfig) -> Result<()> {
         let mut start_version = self.get_start_version(node_config).await?;
+        let mut target_version = self.db_indexer.main_db_reader.ensure_synced_version()?;
+        let mut step_timer = std::time::Instant::now(); // this is just to measure the time used for processing a batch
+                                                        // The Db may opens with data to bootstrap
 
         loop {
-            let start_time: std::time::Instant = std::time::Instant::now();
-            let next_version = self.db_indexer.process_a_batch(start_version)?;
-
-            if next_version == start_version {
-                if let Ok(recv_res) =
-                    tokio::time::timeout(Duration::from_millis(100), self.update_receiver.changed())
-                        .await
-                {
-                    if recv_res.is_err() {
-                        info!("update sender is dropped");
-                        return Ok(());
-                    }
+            // To exclude the case when the main db is not empty and indexer db require bootstrapping
+            if target_version == start_version {
+                match self.update_receiver.changed().await {
+                    Ok(_) => {
+                        (step_timer, target_version) = *self.update_receiver.borrow();
+                    },
+                    Err(e) => {
+                        panic!("Failed to get update from update_receiver: {}", e);
+                    },
                 }
-                continue;
-            };
+            }
+            let next_version = self.db_indexer.process(start_version, target_version)?;
+            INDEXER_DB_LATENCY.set(step_timer.elapsed().as_millis() as i64);
             log_grpc_step(
                 SERVICE_TYPE,
                 IndexerGrpcStep::InternalIndexerDBProcessed,
                 Some(start_version as i64),
                 Some(next_version as i64),
                 None,
                 None,
-                Some(start_time.elapsed().as_secs_f64()),
+                Some(step_timer.elapsed().as_secs_f64()),
```

Comment:
same issue with yesterday, that before it's caught up you already logged the latency, which is not accurate?

Reply:
Each loop now is blocked on notification of write to main db. Previously, each loop is only a batch of all updates. so this should reflect the latency.

Reply:
I see.